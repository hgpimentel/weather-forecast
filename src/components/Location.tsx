import React, { useContext } from "react";
import ColumnContainer from "./ColumnContainer";
import CenteredContainer from "./CenteredContainer";
import LocationContext from "../store/location-context";
import styled from "styled-components";

const CustomColumnContainer = styled(ColumnContainer)`
  padding: 2rem 3rem;

  & > div:last-child {
    margin-top: 1rem;
  }
`;

const Location: React.FC = () => {
  const { currentLocation } = useContext(LocationContext);

  if (!currentLocation) return null;

  return (
    <CenteredContainer>
      <CustomColumnContainer>
        <div>
          <h3>Current location</h3>
        </div>
        {currentLocation.city && currentLocation.country ? (
          <div>{`${currentLocation.city}, ${currentLocation.country}`}</div>
        ) : (
          <div>Browser based</div>
        )}
        <div>Lat: {currentLocation.lat} </div>
        <div>Long: {currentLocation.long} </div>
        <div>
          <button>Change location</button>
        </div>
      </CustomColumnContainer>
    </CenteredContainer>
  );
};

export default Location;
