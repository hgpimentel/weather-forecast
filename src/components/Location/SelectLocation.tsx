import React, { useContext } from "react";
import styled from "styled-components";
import CenteredContainer from "../Container/CenteredContainer";
import ColumnContainer from "../Container/ColumnContainer";
import LocationContext, { Location } from "../../store/location-context";

const CustomCenteredContainer = styled(CenteredContainer)`
  align-items: center;
  justify-content: space-evenly;
  padding: 1rem;
  border-bottom: 2px solid teal;

  & > * {
    flex-grow: 1;
    max-width: 100px;
    display: flex;
    justify-content: center;
  }

  button {
    width: 75px;
    text-align: center;
    margin-top: 0.5rem;
    background-color: teal;
    color: white;
    display: inline-block;
    border-radius: 5px;
    padding: 5px;
    border: none;
    cursor: pointer;

    &:hover {
      background-color: rgba(0, 128, 128, 0.75);
    }
  }
`;

const SelectLocation: React.FC<{ closeModal: () => void }> = ({
  closeModal,
}) => {
  const lsItem = localStorage.getItem("locations");
  const locations = lsItem ? (JSON.parse(lsItem) as Location[]) : null;
  const { onLocationChange } = useContext(LocationContext);

  const onSelect = (location: Location) => {
    onLocationChange(location);
    closeModal();
  };

  return (
    <ColumnContainer>
      <div>
        <h3>Saved locations</h3>
      </div>
      {locations ? (
        <>
          {locations.map((location, index) => {
            return (
              <CustomCenteredContainer key={index}>
                <div>{location.city}</div>
                <div>{location.country}</div>
                <div>{location.lat}</div>
                <div>{location.long}</div>
                <div>
                  <button onClick={() => onSelect(locations[index])}>
                    Select
                  </button>
                </div>
              </CustomCenteredContainer>
            );
          })}
        </>
      ) : (
        <div>
          <h3>No locations found!</h3>
        </div>
      )}
    </ColumnContainer>
  );
};

export default SelectLocation;
