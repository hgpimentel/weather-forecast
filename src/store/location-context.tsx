import React, { useEffect, useState } from "react";

interface Location {
  city: string;
  country: string;
  lat: number;
  long: number;
}

type LocationType = Location | null;

const DefaultLocation: LocationType = {
  country: "Portugal",
  city: "Lisbon",
  lat: 38.7259284,
  long: -9.137382,
};

const LocationContext = React.createContext<{
  currentLocation: LocationType;
  onLocationChange: (location: LocationType) => void;
}>({
  currentLocation: null,
  onLocationChange: () => {},
});

export const LocationContextProvider: React.FC = ({ children }) => {
  const [location, setLocation] = useState<LocationType>(null);

  useEffect(() => {
    setLocation(DefaultLocation);
  }, []);

  return (
    <LocationContext.Provider
      value={{ currentLocation: location, onLocationChange: setLocation }}
    >
      {children}
    </LocationContext.Provider>
  );
};

export default LocationContext;
