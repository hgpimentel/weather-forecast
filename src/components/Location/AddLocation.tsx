import React, { useState, useContext } from "react";
import { ColumnContainer } from "../../components";
import LocationContext, { Location } from "../../store/location-context";
import styled from "styled-components";

const Form = styled.form`
  width: 100%;
  display: flex;
  flex-direction: column;

  span {
    width: 100px;
  }

  & > * {
    margin-bottom: 0.5rem;
    display: flex;
  }

  button {
    width: 100px;
    text-align: center;
    margin-top: 0.5rem;
    background-color: teal;
    color: white;
    display: inline-block;
    padding: 5px;
    border-radius: 5px;
    border: none;
    cursor: pointer;

    &:hover {
      background-color: rgba(0, 128, 128, 0.75);
    }
  }
`;

const AddLocation: React.FC<{ closeModal: () => void }> = ({ closeModal }) => {
  const [city, setCity] = useState<string>("");
  const [country, setCountry] = useState<string>("");
  const [lat, setLat] = useState<number>(0.0);
  const [long, setLong] = useState<number>(0.0);
  const { onLocationChange } = useContext(LocationContext);

  const onSubmit = (e: React.SyntheticEvent) => {
    e.preventDefault();
    const lsItem = localStorage.getItem("locations");
    const locations = lsItem ? (JSON.parse(lsItem) as Location[]) : null;
    const location: Location = {
      city,
      country,
      lat,
      long,
    };

    if (locations) {
      locations.push(location);
      localStorage.setItem("locations", JSON.stringify(locations));
    } else {
      localStorage.setItem("locations", JSON.stringify([location]));
    }

    onLocationChange(location);
    closeModal();
  };

  const canSubmit = city && country && lat && long;

  return (
    <ColumnContainer>
      <div>
        <h3>New location</h3>
      </div>
      <Form>
        <div>
          <span>City</span>
          <input
            type="text"
            id="city"
            name="city"
            value={city}
            onChange={(e) => setCity(e.target.value)}
          />
        </div>
        <div>
          <span>Country</span>
          <input
            type="text"
            id="country"
            name="country"
            value={country}
            onChange={(e) => setCountry(e.target.value)}
          />
        </div>
        <div>
          <span>Latitude</span>
          <input
            type="number"
            id="lat"
            name="lat"
            value={lat}
            onChange={(e) => setLat(e.target.valueAsNumber)}
          />
        </div>
        <div>
          <span>Longitude</span>
          <input
            type="number"
            id="long"
            name="long"
            value={long}
            onChange={(e) => setLong(e.target.valueAsNumber)}
          />
        </div>
        <button type="submit" disabled={!canSubmit} onClick={onSubmit}>
          Add
        </button>
      </Form>
    </ColumnContainer>
  );
};

export default AddLocation;
