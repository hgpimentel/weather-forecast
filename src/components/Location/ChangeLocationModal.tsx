import React from "react";
import ReactDOM from "react-dom";
import styled from "styled-components";
import AddLocation from "./AddLocation";
import SelectLocation from "./SelectLocation";

const Backdrop = styled.div`
  position: fixed;
  top: 0;
  left: 0;
  width: 100%;
  height: 100vh;
  z-index: 10;
  background: rgba(0, 0, 0, 0.75);
  display: flex;
  align-items: center;
  justify-content: center;
`;

const Modal = styled.div`
  z-index: 100;
  background: white;
  width: 50%;
  max-height: 90vh;
  border-radius: 0.5rem;
  overflow: hidden;

  @media (max-width: 991.98px) {
    width: 75%;
  }

  @media (max-width: 767.98px) {
    width: 100%;
    font-size: 14px;
    height: 100vh;
  }

  & > * {
    padding: 0 4rem 2rem;
  }

  header {
    border-bottom: 2px solid teal;
    display: flex;
    justify-content: space-between;
    padding: 0 1rem;

    & > * {
      display: flex;
      align-items: center;
    }

    button {
      background-color: transparent;
      border: none;
      font-weight: bolder;
      cursor: pointer;
    }
  }

  h2 {
    margin: 1rem 0 0.5rem;
  }
`;

const CloseButton: React.FC<{ onCloseModal: () => void }> = ({
  onCloseModal,
}) => {
  return (
    <div>
      <button onClick={onCloseModal}>X Close</button>
    </div>
  );
};

const ChangeLocation: React.FC<{ closeModal: () => void }> = ({
  closeModal,
}) => {
  return (
    <>
      <Backdrop>
        <Modal>
          <header>
            <h2>Change location</h2>
            <CloseButton onCloseModal={closeModal} />
          </header>
          <AddLocation closeModal={closeModal} />
          <SelectLocation closeModal={closeModal} />
        </Modal>
      </Backdrop>
    </>
  );
};

const ChangeLocationModal: React.FC<{ closeModal: () => void }> = ({
  closeModal,
}) => {
  const modal = document.getElementById("modal-root");

  if (!modal) return null;

  return (
    <>
      {ReactDOM.createPortal(<ChangeLocation closeModal={closeModal} />, modal)}
    </>
  );
};

export default ChangeLocationModal;
