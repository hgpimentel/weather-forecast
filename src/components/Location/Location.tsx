import React, { useContext, useState } from "react";
import ColumnContainer from "../Container/ColumnContainer";
import CenteredContainer from "../Container/CenteredContainer";
import LocationContext from "../../store/location-context";
import styled from "styled-components";
import ChangeLocationModal from "./ChangeLocationModal";

const CustomColumnContainer = styled(ColumnContainer)`
  padding: 2rem 3rem;

  & > div:last-child {
    margin-top: 1rem;
  }
`;

const Location: React.FC = () => {
  const { currentLocation } = useContext(LocationContext);

  const [isChangeLocation, setIsChangeLocation] = useState(false);

  const openModal = () => {
    setIsChangeLocation((prev) => !prev);
  };

  if (!currentLocation) return null;

  return (
    <>
      {isChangeLocation && (
        <ChangeLocationModal closeModal={() => setIsChangeLocation(false)} />
      )}
      <CenteredContainer>
        <CustomColumnContainer>
          <div>
            <h3>Current location</h3>
          </div>
          {currentLocation.city && currentLocation.country ? (
            <div>{`${currentLocation.city}, ${currentLocation.country}`}</div>
          ) : (
            <div>Browser based</div>
          )}
          <div>Lat: {currentLocation.lat} </div>
          <div>Long: {currentLocation.long} </div>
          <div>
            <button onClick={openModal}>Change location</button>
          </div>
        </CustomColumnContainer>
      </CenteredContainer>
    </>
  );
};

export default Location;
