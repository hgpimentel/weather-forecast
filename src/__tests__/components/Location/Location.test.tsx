import renderer from "react-test-renderer";
import Location from "../../../components/Location/Location";
import { render, screen } from "@testing-library/react";
import { LocationContextProvider } from "../../../store/location-context";

describe("Location component", () => {
  it("renders correctly", () => {
    const component = renderer.create(<Location />).toJSON();

    expect(component).toMatchSnapshot();
  });

  it("has heading and button enabled", () => {
    render(
      <LocationContextProvider>
        <Location />
      </LocationContextProvider>
    );

    expect(screen.getByRole("heading")).toHaveTextContent("Current:");
    const button = screen.getByRole("button");

    expect(button).toHaveTextContent("Change location");
    expect(button).not.toBeDisabled();
  });
});
