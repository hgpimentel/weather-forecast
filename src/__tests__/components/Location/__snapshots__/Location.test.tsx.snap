// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Location component renders correctly 1`] = `null`;
